core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala

```diff
@@ -0,0 +1,264 @@
+    ZkUtils.makeSurePersistentPathExists(zkClient, SimpleAclAuthorizer.AclZkPath)
+    ZkUtils.makeSurePersistentPathExists(zkClient, SimpleAclAuthorizer.AclChangedZkPath)
+    zkClient.subscribeDataChanges(SimpleAclAuthorizer.AclChangedZkPath, AclChangeListener)
+
+    //we still invalidate the cache every hour in case we missed any watch notifications due to re-connections.
+    scheduler.startup()
+    scheduler.schedule("sync-acls", syncAcls, delay = 0l, period = 1l, unit = TimeUnit.HOURS)
+
+    loadCache
+  }
+
+  override def authorize(session: Session, operation: Operation, resource: Resource): Boolean = {
+    val principal: KafkaPrincipal = new KafkaPrincipal(KafkaPrincipal.USER_TYPE, session.principal.getName)
+    val host = session.host
+
+    trace(s"Authorizer invoked to authorize principal = $principal from host = $host on resource = $resource for operation = $resource")
+
+    if (superUsers.contains(principal)) {
+      debug(s"principal = $principal is a super user, allowing operation without checking acls.")
+      return true
+    }
+
+    val acls: Set[Acl] = getAcls(resource)
+    if (acls.isEmpty) {
+      debug(s"No acl found for resource $resource , authorized = $shouldAllowEveryoneIfNoAclIsFound")
+      return shouldAllowEveryoneIfNoAclIsFound
+    }
+
+    //first check if there is any Deny acl match that would disallow this operation.
+    if (aclMatch(session, operation, resource, principal, host, Deny, acls))
+      return false
+
+    /**
+     * if principal is allowed to read or write we allow describe by default, the reverse does not apply to Deny.
+     */
+    val ops: Set[Operation] = if (Describe.equals(operation))
+      Set[Operation](operation, Read, Write)
+    else
+      Set[Operation](operation)
+
+    //now check if there is any allow acl that will allow this operation.
+    if (ops.exists(operation => aclMatch(session, operation, resource, principal, host, Allow, acls)))
+      return true
+
+    //We have some acls defined and they do not specify any allow ACL for the current session, reject request.
+    debug("principal = %s from host = %s is not allowed to perform operation = %s  on resource = %s as no explicit acls were defined to allow the operation".format(principal, host, operation, resource))
+    false
+  }
+
+  private def aclMatch(session: Session, operations: Operation, resource: Resource, principal: KafkaPrincipal, host: String, permissionType: PermissionType, acls: Set[Acl]): Boolean = {
+    for (acl: Acl <- acls) {
+      if (acl.permissionType.equals(permissionType)
+        && (acl.principal.equals(principal) || acl.principal.equals(Acl.WildCardPrincipal))
+        && (operations.equals(acl.operation) || acl.operation.equals(All))
+        && (acl.host.contains(host) || acl.host.contains(Acl.WildCardHost))) {
+        debug(s"operation = $operations on resource = $resource from host = $host is $permissionType.name based on acl = $acl")
+        return true
```

Comment:
the original version is more readable to me, may be its just my java brain. I have assumed scala idioms make it more redable with your approach so changed it as you suggested. Still had to use a return inside the map as acl is not an Option but type ACL.